### DISCUSSION TITLE
How do I associate multiple wallet addresses to a next.id avatar DID?

### DISCUSSION THREAD

**Discussion by javaspeak** [NextDotID/proof_server#105] [Q&A]
How do I associate multiple wallet addresses to a next.id avatar DID without having access to the private keys.  i.e. do this only using metamask wallets?

**Comment by javaspeak** [comment]
BACKROUND FOR CONTEXT:

(Note the answer is further down)

Adding an address looks like this:

POST: https://proof-service.nextnext.id/v1/proof

Payload:

```
{
    "action": "create",
    "platform": "ethereum",
    "identity": "0x8dd1Cc73D574608BA7E27C485dA4E3645C62FBf6",
    "public_key": "0x046ca64e70760e7635e10161095693abb08e9548ff84bed952fc08ce3ccc196736383d1673f699625e78080b9072d433bbac69e8f81d3430d9746615b69d9e9924",
    "extra": {
        "wallet_signature": "gQWh59jDC/ryj1F9+2mV8oMk1n5sPfsab0E3G0v4EWcGacZmuIymDtEm6dn6OupjLChSKBAY5EsGAU0qK96e4Rw=",
        "signature": "gQWh59jDC/ryj1F9+2mV8oMk1n5sPfsab0E3G0v4EWcGacZmuIymDtEm6dn6OupjLChSKBAY5EsGAU0qK96e4Rw="
    },
    "uuid": "7b6b7d3f-5004-4469-b727-42ddd1f3a46f",
    "created_at": "1705862170"
}
```

Notice that signature comes from the wallet that signed the avatar and that the wallet_signature is the wallet address you want to add.

In this example I have already shown how to do it when both the wallet address signing the avatar and the wallet address I want to asoociate with my avatar DID are the same:

https://github.com/spotadev/next-id-twitter-wallet

However I want to do it when they are different and without the user handling private keys.  i.e. only want the wallets involved to managing private keys and doing the signing.


ANSWER:  

Meta Mask can have multiple addresses in it.  So maybe the user could import all the addresses they want to associate with their DID into metamask.

Then the user could connect to each of the wallets it wants associated with its avatar DID, in turn and create the signatures.  The signatures could be stored in browser storage.

Finally the user could switch to the wallet which is managing the avatar and dig out all the other signatures from storage and send them to the proof service.

**Comment by nykma** [comment]
> without having access to the private keys

Using wallet containers like MetaMask is still considered as "having access". 

Since we haven't impl a dedicated Avatar private key management (will be impl in MPC way later). For now: 

- we suggest apps / projects to guide user to create a new "wallet" in MetaMask and label it as something like "Avatar"
- Or, guide user to use a wallet container which has the concept of "avatar", like [Mask Network](https://mask.io/download-links)
  - MaskNetwork browser plugin is built on the concept "Persona" , which is almost the same design goal as our "avatar".
